### PR TITLE
Add auto remediation

### PR DESCRIPTION
**Pull request by pauloconnor**
This is a bit of a first pass/WIP of auto remediation.

Adding 3 lines to the monitoring_check definition:
use_remediation
remediation_action
remediation_retries

will wrap the check in a script. The script will run the check as normal. If it passes, the wrapper will pass the check output to stdout, and pass the check exit code correct. If the check fails, it will run the remediation script, recording that it has run it in a file called /tmp/$name. The number of times the remediation action will run can be set. Only once this number of attempts has been reached will the check properly fail, and page/ticket as appropriate.

It's a bit janky - some of the bash output could be cleaned up, and it may make more sense to move this to Ruby, but it works for now.

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -220,10 +233,26 @@
     $irc_channel_array = $team_hash[$team]['notifications_irc_channel']
   }
 
+  if $use_remediation {
+    file { "${monitoring_check::params::etc_dir}/plugins/remediation.sh":
```
Comment (somic) [inline manifests/init.pp]: If there are 2 or more monitoring_checks on a host that use remediation, wouldn't it cause duplication declaration of this File?

Comment (pauloconnor) [inline manifests/init.pp]: Ah, I was wondering where I'd be tripped up about that. Good catch.

File: files/remediation.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  r)
+    # Number of times to retry the action
+    RETRIES=$OPTARG
+    ;;
+  esac
+done
+
+CHECK_OUTPUT=$($CHECK)
+CHECK_EXITCODE=$?
+if [ $CHECK_EXITCODE -ne 0 ]; then
```
Comment (somic) [inline files/remediation.sh]: Do we remediate on all non-OK return codes (warning, critical, unknown)? I think we could start calling remediation only on criticals ($CHECK_EXITCODE -ne 2) and then iterate if necessary.

Comment (somic) [inline files/remediation.sh]: So it would be $CHECK_EXITCODE -eq 2

File: files/remediation.sh [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+echo $1
+while getopts "n:c:a:r:" arg; do
+  case $arg in
+  n)
+    # The name of the check we're running
+    NAME=$OPTARG
+    ;;
+  c)
+    # The check that we're running
+    CHECK=$OPTARG
```
Comment (somic) [inline files/remediation.sh]: Some crazy escaping could lead to unexpected results here and below where $CHECK is used.

File: manifests/init.pp [hunk 4]
```diff
@@ -103,6 +103,15 @@
 # These will override any parameters configured by the wrapper.
 # Defaults to an empty hash.
 #
+# [*use_remediation*]
```
Comment (solarkennedy) [inline manifests/init.pp]: What if instead having another var, just do

    if $remediation_action != under {

?

File: manifests/init.pp [hunk 5]
```diff
@@ -103,6 +103,15 @@
...
+# Should a script run automatically when a check fails?
+#
+# [*remediation_action*]
+# Which script should run when a check fails?
+#
+# [*remediation_retries*]
```
Comment (solarkennedy) [inline manifests/init.pp]: I'm concerned about the complexity with retries. Do we need it?

The checks run over and over anyway on every interval.

File: manifests/init.pp (hunk 4 above)
Comment (pauloconnor) [inline manifests/init.pp]: I was considering that alright. I quite like the ability to toggle remediation on and off without having to change the action name.

File: manifests/init.pp (hunk 5 above)
Comment (pauloconnor) [inline manifests/init.pp]: If we don't have the retries, the remediation step will run continually, and we will never get paged. I see this as a method of attempting a step that would be done manually (say, restarting a service), and if that fails to fix the problem, then we should be paged.

File: files/remediation.sh (hunk 3 above)
Comment (pauloconnor) [inline files/remediation.sh]: I'm not sure I follow. Do you have an example of what you're thinking?

File: manifests/init.pp (hunk 5 above)
Comment (solarkennedy) [inline manifests/init.pp]: Gotcha.

File: files/remediation.sh (hunk 3 above)
Comment (somic) [inline files/remediation.sh]: I don't have an example but we are passing entire command line (with args and whitespaces and backslashes and stuff) to this. All I am saying is this *could* give unexpected results.

Comment (pauloconnor) [inline files/remediation.sh]: True, but I would hope that people would test this carefully :)

If there's a nicer way of doing this, I'd love to know it.

Comment (pauloconnor) [inline files/remediation.sh]: So, puppet stdlib has shell_escape, which will safely escape these commands.

**Comment by pauloconnor** [comment]
@somic @solarkennedy Anything else needed on this PR? I've a couple of ideas for iterating on this, and I'll probably rewrite it in Ruby at some point, but I think it's solid enough for now.

**Comment by somic** [comment]
ship it

**Pull request merged**